### ISSUE TITLE
Pro version/Pundit integration/Mutation: How to get the object, not the mutation?

### ISSUE THREAD

**Issue by ralph** [rmosolgo/graphql-ruby#2367]
Hi,

I bought the pro version and set up pundit integration according to [this guide](https://graphql-ruby.org/authorization/pundit_integration#mutation-level-roles), but I'm having trouble getting mutations to work. Consider this mutation class:

```ruby
module Mutations
  class AcceptAttendee < Mutations::BaseMutation
    pundit_role :accept

    argument :id, ID, required: true # , loads: Types::AttendeeType
    field :attendee, Types::AttendeeType, null: true
    field :errors, [String], null: false

    def resolve(id:)
      attendee = Attendee.find(id)
      if attendee.accept
        {
          attendee: attendee,
          errors: []
        }
      else
        {
          attendee: nil,
          errors: attendee.errors.full_messages
        }
      end
    end

    def self.policy_class
      self.const_get(:Policy)
    end

    class Policy < ::ApplicationPolicy
      def accept?
        debugger
        true
      end
    end
  end
end
```

When I run this mutation, and the ruby process stops at the debugger line, I can inspect the record that is passed to the policy class. And much to my surprise, it's an instance of the `AcceptAttendee` mutation class:

```
(byebug) record
mutation AcceptAttendee
```

The mutation works as expected when I remove the debugger, but instead of passing `true`, I want to implement a real check whether the `current_user` has access to the `Attendee` record. So in this position, instead of an instance of the `AcceptAttendee` class, I need an instance of the `Attendee` class. Is that understandable? Am I doing something wrong here?

I also tried adding the `load` option to the `argument` macro, but this throws the following error:

```
NotImplementedError (MySchema.object_from_id(node_id, ctx) must be implemented to load by ID (tried to load from id `26713`)):
```

I'm not sure how to implement this, also I expected this to work out of the box given I have specified an `AttendeeType` that works well in queries (and in the mutation once I return `true` from the `accept?` method.

Anyway, I'm stuck and appreciate any help. Thank you in advance.

**Comment by rmosolgo** [comment]
Hi! Thanks for the detailed writeup and sorry about the confusion.

I think the configuration you want is like this: 

```ruby 
argument :id, ID, required: true, loads: Types::AttendeeType, 
  pundit_role: :accept?,
  pundit_policy_class: "Mutations::AcceptAttendee::Policy"
```

That will load the attendee by ID and use the mutation policy's `def accept?`  to see whether the current user can accept the attendee or not. 

BUT, as you noticed, automatic loading by ID requires the ID hooks to be implemented. 

If you're using Rails, an easy way to do it is with GlobalID (https://github.com/rails/globalid):

```ruby
class MySchema < GraphQL::Schema 
  # @return [String] generate a unique ID for `object` using Rails's GlobalID library
  def self.id_from_object(object, type, context)
    object.to_global_id.to_s 
  end
  # @return [Object] Load the object specified by `unique_id`, using Rails's GlobalID library
  def self.object_from_id(unique_id, context)
    GlobalID::Locator.locate(unique_id)
  end
end 
```


This is pretty similar to what Shopify did in their recent Storefront API release. You can see some examples at: https://help.shopify.com/en/api/graphql-admin-api/migrating-from-rest#translating-from-rest-ids-to-graphql-ids

Does that help? Let me know what you find!

**Comment by ralph** [comment]
Hi @rmosolgo, I did as advised, but for some reason, the `record` inside the policy is always `nil`. Also the Rails logs don't show any SQL code of the Attendee being loaded.

After studying the source, I also tried a different version by specifying a `load_attendee` method, which works if `argument` does not have a `pundit_role` specified, but also does not work once I specify a pundit role for the argument.

So this works (loads Attendee):

```
module Mutations
  class AcceptAttendee < Mutations::BaseMutation
    pundit_role nil

    argument :attendee_id, ID, required: true, loads: Types::AttendeeType
    field :attendee, Types::AttendeeType, null: true
    field :errors, [String], null: false

    def load_attendee(attendee_id)
      Attendee.find(attendee_id)
    end

    def resolve(attendee:)
      # attendee is loaded correctly
      policy = pundit_policy(context, attendee)
      …
  end
end
```

But this does not work:

```
module Mutations
  class AcceptAttendee < Mutations::BaseMutation
    pundit_role nil

    argument :attendee_id, ID, required: true, loads: Types::AttendeeType, pundit_role: :accept, pundit_policy_class: 'AttendeePolicy'
    field :attendee, Types::AttendeeType, null: true
    field :errors, [String], null: false

    def load_attendee(attendee_id)
      Attendee.find(attendee_id)
    end

    def resolve(attendee:)
      # we don't end up here, because record is nil inside the policy.
      …
  end
end
```

The result is the same whether I use the GlobalID/schema or the `load_attendee` approach.

Do you have any more ideas?

**Comment by rmosolgo** [comment]
🤔 Hmmm, sorry that didn't work! I just spent some time with the GraphQL-Pro test suite to see if I could find a configuration that caused the unit tests to fail, but they seem to work ok. So I must be missing something! 

Could you share your versions of GraphQL-Ruby and GraphQL-Pro? If you're not on the latest, you could try updating. 

Also, could you share the debug code you're using to determine that the `record` is nil? Maybe that would help me isolate the problem in my tests. 

It's so odd because I _expect_ `#load_attendee` to come _before_ any authorization checks on the argument, so I can't think of how `pundit_role:` would affect argument loading. If you add _just_ `pundit_policy_class:`, does that also cause `#load_attendee` to be skipped? (Just trying to find something that might be a clue!)

**Comment by ralph** [comment]
Thank you for your help, @rmosolgo. In the mean time, our team refactored the existing queries & mutations to match the Relay standard, especially for pagination support. We are all pretty new to GraphQL and hat to experiment and read a little before settling on this. While doing the refactoring, we also switched to GlobalID for object identification. After the refactoring was done, I tried your approach once again, and this time it worked. I don't know if it's because we're using `RelayClassicMutation` as a base for `BaseMutation` now or because my first try with GlobalID was bugged, but the important part is: It works now! :) So thank you very much.
